### PR TITLE
Fix #16 - gzopen() / gzopen64() - 32 bit builds of Ubuntu 14.04

### LINKED ISSUE #16: gzopen()  / gzopen64()
On 32bit builds of Ubuntu 14.04LTS (PHP 5.5.9), three functions are created with the wrong names.
[See one of many bug reports.](https://bugs.launchpad.net/ubuntu/+source/php5/+bug/1315888)
There seems to be no plans to fix this in Ubuntu, so users of this LTS release cannot use the GZIP library.

Is there a place in this package for the necessary polyfills?

```php
if (!function_exists('gzopen') && function_exists('gzopen64')) {
	function gzopen($filename , $mode = 'r', $use_include_path = 0 ) {
		return gzopen64($filename, $mode, $use_include_path);
	}
}
if (!function_exists('gzseek') && function_exists('gzseek64')) {
	function gzseek($zp , $offset, $whence = SEEK_SET) {
		return gzseek64($zp, $offset, $whence);
	}
}
if (!function_exists('gztell') && function_exists('gztell64')) {
	function gztell($zp) {
		return gztell64($zp);
	}
}
```

### PR DESCRIPTION
**Pull request by fisharebest**
Fix for #16

### CODE REVIEW

File: src/Php56/bootstrap.php [hunk 1]
```diff
@@ -21,4 +21,16 @@ function hash_equals($knownString, $userInput) { return p\Php56::hash_equals($kn
...
+    // Missing functions in PHP 5.5.9 - affects 32 bit builds of Ubuntu 14.04LTS
+    // See https://bugs.launchpad.net/ubuntu/+source/php5/+bug/1315888
+    if (!function_exists('gzopen') && function_exists('gzopen64')) {
+        function gzopen($filename , $mode = 'r', $use_include_path = 0 ) { return gzopen64($filename, $mode, $use_include_path); }
+    }
+    if (!function_exists('gzseek') && function_exists('gzseek64')) {
+        function gzseek($zp , $offset, $whence = SEEK_SET) { return gzseek64($zp, $offset, $whence); }
+    }
+    if (!function_exists('gztell') && function_exists('gztell64')) {
+        function gztell($zp) { return gztell64($zp); }
```
Comment (keradus) [inline src/Php56/bootstrap.php]: `return p\Phy56::gztell64($zp);` ?

**Comment by keradus** [comment]
adding some tests would be awesome!

**Comment by fisharebest** [comment]
The travis failure is not in my code ;-)

It looks like php-7 is using phpunit-5, but perhaps the tests were written for phpunit-4

**Comment by keradus** [comment]
ref https://github.com/symfony/polyfill/pull/15 ( and https://github.com/symfony/polyfill/pull/15#issuecomment-162552595 )

File: src/Php56/bootstrap.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: i'm ok with not duplicating the function as static method in the Php56 class (reduces boiler plates)

File: src/Php56/bootstrap.php [hunk 2]
```diff
@@ -21,4 +21,16 @@ function hash_equals($knownString, $userInput) { return p\Php56::hash_equals($kn
 
         function ldap_escape($subject, $ignore = '', $flags = 0) { return p\Php56::ldap_escape($subject, $ignore, $flags); }
     }
+
+    // Missing functions in PHP 5.5.9 - affects 32 bit builds of Ubuntu 14.04LTS
+    // See https://bugs.launchpad.net/ubuntu/+source/php5/+bug/1315888
+    if (!function_exists('gzopen') && function_exists('gzopen64')) {
+        function gzopen($filename , $mode = 'r', $use_include_path = 0 ) { return gzopen64($filename, $mode, $use_include_path); }
```
Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: $mode doesn't default to 'r' per the doc (http://php.net/gzopen)
trailing space before `)`

File: src/Php56/bootstrap.php (hunk 1 above)
Comment (stof) [inline src/Php56/bootstrap.php]: why that ?

**Comment by stof** [comment]
The Travis failure is because of a BC break in PHPUnit 5.1

File: src/Php56/bootstrap.php (hunk 2 above)
Comment (stof) [inline src/Php56/bootstrap.php]: and shouldn't ``$use_include_path`` be a boolean instead ?

Comment (stof) [inline src/Php56/bootstrap.php]: ah no, they use a crappy int API instead of a proper boolean

File: src/Php56/bootstrap.php (hunk 1 above)
Comment (keradus) [inline src/Php56/bootstrap.php]: more consistency and easier to test then moved into `Php56`

Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: I wonder @keradus how would you test that? We'd need the broken ubuntu isn't it?

**Comment by fisharebest** [comment]
> adding some tests would be awesome!

Since we are working around a platform-specific problem, I'm not sure how we create a meaningful test?

To trigger the code, we'd need to create a situation where the function `gzopen64()` does not exist.  Since we cannot "unload" PHP libraries at runtime, I don't see how to do this.

Besides, none of the other "thin wrappers" have unit tests.

FYI, I've tested the code on a VM...

**Comment by fisharebest** [comment]
> $mode doesn't default to 'r'

Oops.  My bad.  Fixed

File: src/Php56/bootstrap.php [hunk 3]
```diff
@@ -21,4 +21,16 @@ function hash_equals($knownString, $userInput) { return p\Php56::hash_equals($kn
...
         function ldap_escape($subject, $ignore = '', $flags = 0) { return p\Php56::ldap_escape($subject, $ignore, $flags); }
     }
+
+    // Missing functions in PHP 5.5.9 - affects 32 bit builds of Ubuntu 14.04LTS
+    // See https://bugs.launchpad.net/ubuntu/+source/php5/+bug/1315888
+    if (!function_exists('gzopen') && function_exists('gzopen64')) {
+        function gzopen($filename , $mode, $use_include_path = 0) { return gzopen64($filename, $mode, $use_include_path); }
+    }
+    if (!function_exists('gzseek') && function_exists('gzseek64')) {
+        function gzseek($zp , $offset, $whence = SEEK_SET) { return gzseek64($zp, $offset, $whence); }
```
Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: extra space before first comma (same above)

File: src/Php56/bootstrap.php [hunk 4]
```diff
@@ -21,4 +21,16 @@ function hash_equals($knownString, $userInput) { return p\Php56::hash_equals($kn
 
         function ldap_escape($subject, $ignore = '', $flags = 0) { return p\Php56::ldap_escape($subject, $ignore, $flags); }
     }
+
+    // Missing functions in PHP 5.5.9 - affects 32 bit builds of Ubuntu 14.04LTS
+    // See https://bugs.launchpad.net/ubuntu/+source/php5/+bug/1315888
+    if (!function_exists('gzopen') && function_exists('gzopen64')) {
+        function gzopen($filename , $mode, $use_include_path = 0) { return gzopen64($filename, $mode, $use_include_path); }
```
Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: extra space before first comma

Comment (fisharebest) [inline src/Php56/bootstrap.php]: > extra space before first comma

Removed.

File: src/Php56/bootstrap.php [hunk 5]
```diff
@@ -21,4 +21,16 @@ function hash_equals($knownString, $userInput) { return p\Php56::hash_equals($kn
 
         function ldap_escape($subject, $ignore = '', $flags = 0) { return p\Php56::ldap_escape($subject, $ignore, $flags); }
     }
+
+    // Missing functions in PHP 5.5.9 - affects 32 bit builds of Ubuntu 14.04LTS
```
Comment (nicolas-grekas) [inline src/Php56/bootstrap.php]: I propose to enclose these checks in a `if (8 > PHP_INT_SIZE) {` (and removing it from this `PHP_VERSION_ID < 50600` block) to get a few more CPU cycles on 64b setups (which are the most common)

Comment (fisharebest) [inline src/Php56/bootstrap.php]: Just powered up my 14.04 / 32bit virtual machine...

`if (PHP_VERSION_ID === 50509 && PHP_INT_SIZE === 4) { ... }` works nicely for me.

**Comment by nicolas-grekas** [comment]
:+1:

File: src/Php56/bootstrap.php (hunk 5 above)
Comment (stof) [inline src/Php56/bootstrap.php]: Could it affect a few other versions of PHP shipped in other distributions than Ubuntu 14.04 ?

Comment (fisharebest) [inline src/Php56/bootstrap.php]: @stof - not to my knowledge.  All google searches and bug reports point to just the 32bit build of 14.04

**Comment by nicolas-grekas** [comment]
Thank you @fisharebest.

**Pull request merged**